### PR TITLE
feat: Apply coupons

### PR DESCRIPTION
**Pull request by azhur**
## Description
Apply coupons to subscriptions and invoices.

Todo

- [x] invoicing

## Checklist
- [ ] The code follows the project's coding conventions and style [guidelines](https://doc.rust-lang.org/1.0.0/style/README.html).
- [ ] All tests related to the changes have passed successfully.
- [ ] Documentation has been updated to reflect the changes (if applicable).
- [ ] All new and existing unit tests have passed.
- [ ] I have self-reviewed my code and ensured its quality.
- [ ] I have added/updated necessary comments to aid understanding.

### CODE REVIEW

File: modules/meteroid/src/api/coupons/service.rs [hunk 1]
```diff
@@ -54,8 +54,8 @@ impl CouponsService for CouponsServiceComponents {
             expires_at: req.expires_at.map(chrono_from_timestamp).transpose()?,
             redemption_limit: req.redemption_limit,
             tenant_id,
-            recurring_value: -1, // todo fixme later
-            reusable: false,     // todo fixme later
+            recurring_value: None, // todo fixme later
+            reusable: false,       // todo fixme later
```
Comment (gaspb) [inline modules/meteroid/src/api/coupons/service.rs:58]: this has been implemented no ?

File: modules/meteroid/crates/meteroid-store/src/domain/invoices.rs [hunk 2]
```diff
@@ -262,6 +289,63 @@ impl InvoiceTotals {
...
+        invoice_currency: &str,
+        coupons: &[AppliedCouponDetailed],
+    ) -> AppliedCouponsDiscount {
+        let applicable_coupons: Vec<&AppliedCouponDetailed> = coupons
+            .iter()
+            .filter(|x| x.is_invoice_applicable())
+            .sorted_by_key(|x| x.applied_coupon.created_at)
+            .collect::<Vec<_>>();
+
+        let mut applied_coupons: BTreeMap<Uuid, i64> = BTreeMap::new();
```
Comment (gaspb) [inline modules/meteroid/crates/meteroid-store/src/domain/invoices.rs:307]: so currently coupons are ordered by ids, right ? (that's fine in v1, it's just to understand)

File: modules/meteroid/crates/meteroid-store/src/domain/invoices.rs [hunk 3]
```diff
@@ -262,6 +289,63 @@ impl InvoiceTotals {
...
+            }
+            let discount = match &applicable_coupon.coupon.discount {
+                CouponDiscount::Percentage(percentage) => {
+                    subtotal_subunits * percentage / Decimal::ONE_HUNDRED
+                }
+                CouponDiscount::Fixed { amount, currency } => {
+                    // todo currency conversion
+                    if currency != invoice_currency {
+                        continue;
+                    }
```
Comment (gaspb) [inline modules/meteroid/crates/meteroid-store/src/domain/invoices.rs:323]: can we check it & reject when applying the coupon to the subscription for transparency ?

**Review by gaspb — APPROVED** [review]

File: modules/meteroid/crates/meteroid-store/src/domain/invoices.rs (hunk 2 above)
Comment (azhur) [reply modules/meteroid/crates/meteroid-store/src/domain/invoices.rs:307]: omg, I misused here BTreeMap, I will switch to either Vec or find a map that preserves the order

File: modules/meteroid/src/api/coupons/service.rs (hunk 1 above)
Comment (azhur) [reply modules/meteroid/src/api/coupons/service.rs:58]: yes, I need to add them to proto so then I can map them here 👍

Comment (azhur) [reply modules/meteroid/src/api/coupons/service.rs]: fixed

File: modules/meteroid/crates/meteroid-store/src/domain/invoices.rs (hunk 2 above)
Comment (azhur) [reply modules/meteroid/crates/meteroid-store/src/domain/invoices.rs]: fixed

File: modules/meteroid/crates/meteroid-store/src/domain/invoices.rs (hunk 3 above)
Comment (azhur) [reply modules/meteroid/crates/meteroid-store/src/domain/invoices.rs:322]: fixed

**Pull request merged**